walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -39,46 +41,90 @@
+ 
+        //Get blocked and available times for optional people
+        for(Event event: events){
+            for(String attendee: request.getOptionalAttendees()){
+                if(event.getAttendees().contains(attendee)){
+                    bookedOptionalTimes.add(event.getWhen());
+                    break;
+                }
+            }
+        }
+        List<TimeRange> availableOptionalTimes = getAvailableSlots(bookedOptionalTimes, request.getDuration());
 
-        if(bookedTimes.size() == 0){
-            slots.add(TimeRange.WHOLE_DAY);
+        // If no required attendee is available, then return no time slot available.
+        if(availableTimes.size() == 0){
             return slots;
         }
-
-        //Sort the booked times by start of day
-        Collections.sort(bookedTimes, TimeRange.ORDER_BY_START);
-
-        //Merge overlapping or nested timeslots into a single chunk of time
-        int i = 0;
-        while(i < bookedTimes.size() - 1){
-            if(bookedTimes.get(i).overlaps(bookedTimes.get(i+1)) || bookedTimes.get(i).contains(bookedTimes.get(i+1))){
-                int endTime = Math.max(bookedTimes.get(i).end(), bookedTimes.get(i+1).end());
-                bookedTimes.set(i, TimeRange.fromStartEnd(bookedTimes.get(i).start(), endTime, false));
-                bookedTimes.remove(i+1);
-            }else{
-                i++;
+ 
+        // Compare the timeslots available for required attendees and optional attendees and return the intersection.
+        for(int i = 0; i < availableTimes.size(); i++){
+            for(int j = 0; j < availableOptionalTimes.size(); j++){
+                if(availableTimes.get(i).end() < availableOptionalTimes.get(j).start()){
+                    break;
+                }
+                else if(availableTimes.get(i).equals(availableOptionalTimes.get(j))){
+                    slots.add(TimeRange.fromStartEnd(availableOptionalTimes.get(i).start(), availableOptionalTimes.get(i).end(),false)); 
+                }else if(availableTimes.get(i).contains(availableOptionalTimes.get(j))){
+                    slots.add(TimeRange.fromStartEnd(availableOptionalTimes.get(j).start(), availableOptionalTimes.get(j).end(),false));
+                }else if(availableOptionalTimes.get(j).contains(availableTimes.get(i))){
+                    slots.add(TimeRange.fromStartEnd(availableTimes.get(i).start(), availableTimes.get(i).end(),false));
+                }
             }
         }
 
-        //Get everthing in between the booked time slots as eligible slots for the request
-        long duration = request.getDuration();
-        
-        if((bookedTimes.get(0).start() - TimeRange.START_OF_DAY) >= duration){
-            slots.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, bookedTimes.get(0).start(), false));
+        // If no available overlap is available, return the available times for required attendees.
+        if(slots.size() == 0 && availableOptionalTimes.size() != 0){
```

Comment:
shouldn't you check if (availableOptionalTimes.size() == 0) before the available times intersection logic so that you can immediately return just availableTimes?

Reply:
This is mostly true. The only case it would fail would be if there are no required attendees, and the optional attendees have no viable gaps in their schedule. In such a case, we would want to return an empty list. Checking if availableOptionalTimes.size() is 0 and returning availableTimes would result in us getting a range for the whole day, but we want an empty list

Reply:
Also, the last if check (where slot.size() is 0 and availableOptionalTimes.size() is 0) is necessary because if we have no viable overlap in the optional and required slots, then I should just revert and return the mandatory time-slots

Reply:
so what I'm suggesting is:

List<TimeRange> availableOptionalTimes = getAvailableSlots(bookedOptionalTimes, request.getDuration());
if availableOptionalTimes.size() == 0 {
  if requiredAttendees().size() == 0 {
    return emptyList;
  }
  return availableTimes;
}

..
for (){
 for() {

 }
}
..
if (slots.size() == 0) {
  return availableTimes
}

Will this work?

Reply:
Yes! That would work. Let me implement

Reply:
I don't think it would work. If there are no timeslots available for optional attendees, then availableOptionalTimes.size() would be 0, and availableTimes would contain a range for the whole day. In the if statement above, we would return the range for the whole day, but we want an empty list


This might work tho: 

<pre><code>
if (availableOptionalTimes.size() == 0){
 if(availableTimes.get(0).duration == TimeRange.WHOLE_DAY){
    return emptylist;
 }
 else{
    return availableTimes;
 }
}</code></pre>

Reply:
Sure go ahead.

Reply:
Resolved